src/lib/y2partitioner/dialogs/partition_size.rb

```diff
@@ -10,143 +13,349 @@ module Dialogs
+        @disk_name = disk_name
         @ptemplate = ptemplate
-        @slots = slots
+        @regions = regions
       end
 
+      # @macro seeDialog
       def title
         # dialog title
-        Yast::Builtins.sformat(_("Add Partition on %1"), @disk.name)
+        Yast::Builtins.sformat(_("Add Partition on %1"), @disk_name)
       end
 
+      # @macro seeDialog
       def contents
-        VBox(SizeWidget.new(@disk, @ptemplate, @slots))
+        HVSquash(SizeWidget.new(@ptemplate, @regions))
       end
 
-      # Choose partition size: maximum, size based, start-end based.
-      class SizeWidget < CWM::CustomWidget
-        include Yast
-        include Yast::UIShortcuts
+      # Like CWM::RadioButtons but each RB has a subordinate indented widget.
+      # This is kind of like Pager, but all Pages being visible at once,
+      # and enabled/disabled.
+      # Besides {#items} you need to define also {#widgets}.
+      class ControllerRadioButtons < CWM::CustomWidget
+        def initialize
+          self.handle_all_events = true
+        end
+
+        # @return [WidgetTerm]
+        def contents
+          Frame(
+            label,
+            HBox(
+              HSpacing(hspacing),
+              RadioButtonGroup(Id(widget_id), buttons_with_widgets),
+              HSpacing(hspacing)
+            )
+          )
+        end
+
+        # @return [Numeric] margin at both sides of the options list
+        def hspacing
+          1.45
+        end
+
+        # @return [Numeric] margin above, between, and below the options
+        def vspacing
+          0.45
+        end
+
+        # @return [Array<Array(id,String)>]
+        abstract_method :items
+
+        # FIXME: allow {WidgetTerm}
+        # @return [Array<AbstractWidget>]
+        abstract_method :widgets
+
+        # @param event [Hash] UI event
+        def handle(event)
+          eid = event["ID"]
+          ids.zip(widgets).each do |id, widget|
+            if id == eid
+              widget.enable
+            else
+              widget.disable
+            end
+          end
+          nil
+        end
+
+        # Get the currently selected radio button from the UI
+        def value
+          Yast::UI.QueryWidget(Id(widget_id), :CurrentButton)
+        end
+
+        # Tell the UI to change the currently selected radio button
+        def value=(val)
+          Yast::UI.ChangeWidget(Id(widget_id), :CurrentButton, val)
+        end
+
+        # @return [AbstractWidget] the widget corresponding
+        #   to the currently selected option
+        def current_widget
+          idx = ids.index(value)
+          widgets[idx]
+        end
+
+      private
+
+        # @return [Array<id>]
+        def ids
+          @ids ||= items.map(&:first)
+        end
 
-        def initialize(disk, ptemplate, slots)
+        def buttons_with_widgets
+          items = self.items
+          widgets = self.widgets
+          raise ArgumentError unless items.size == widgets.size
```

Comment:
please specify here message in Argument error, especially when problem is not in arguments.